### ISSUE TITLE
Align DistributionArray container surface with numpy/jax: `__iter__`, `.size`, doc fix

### ISSUE THREAD

**Issue by jhuggins** [TARPS-group/prob-pipe#178]
## Summary

`DistributionArray` is `Array[Distribution]` — its container surface should match the numpy / jax ecosystem so users' array-shaped intuitions transfer. Two places it currently diverges, plus one stale doc.

## Current behaviour

```python
da.batch_shape        # (2, 3)
len(da)               # 2                  ✓ leading axis (matches numpy)
list(da)              # 6 scalar cells     ✗ flat iteration, unlike numpy
da.size               # AttributeError     ✗ no canonical total-cells accessor
```

| Op | numpy / jax | DistributionArray | Aligned? |
|---|---|---|---|
| `len(x)` | leading axis (`shape[0]`) | leading axis (`_batch_shape[0]`) | ✓ |
| `x.size` | `prod(shape)` | does not exist | ✗ |
| `for cell in x` | yields leading-axis slices (sub-arrays for ndim≥2, scalars for ndim==1) | yields all `prod(batch_shape)` cells flat | ✗ |

The flat-iteration semantic is a historical PyTorch-style "treat as flat" behaviour from before `DistributionArray` supported multi-d `batch_shape`. Now that multi-d is the documented contract (see `STYLE_GUIDE.md` §1.11 and the `ArrayRandomFunction` shape table), the iteration semantic should follow numpy.

## Proposed cleanup

1. **`DistributionArray.__iter__` iterates the leading axis.** For a `DistributionArray` of `batch_shape=(2, 3)`:
   - 1-D case (`batch_shape=(n,)`): yield each scalar `Distribution` cell. *(unchanged from today.)*
   - Multi-d case: yield `DistributionArray` sub-arrays of shape `batch_shape[1:]`, mirroring `iter(np.zeros((2, 3)))` yielding two `(3,)`-shaped views.

   Today's flat behaviour stays accessible via `da.components` (already the documented escape hatch — flat tuple in row-major order) or `for i in range(prod(da.batch_shape)): da._flat_component(i)`.

2. **Add `DistributionArray.size`.** Returns `prod(self._batch_shape)`. Migration: callers that want the total cell count can write `da.size` instead of `prod(da.batch_shape)` or the now-removed `da.n`. Matches `np.ndarray.size` / `jax.Array.size`.

3. **Fix `CONTRIBUTING.md:391-392`.** The framework-hierarchy rule 3 says `len(da) == prod(da.batch_shape)`, which contradicts both the implementation and the numpy convention. Replace with the wording already used in `STYLE_GUIDE.md` §1.11 (`len(da)` is the leading-axis size; `prod(da.batch_shape)` is the total cell count).

## Migration impact

The behaviour change is `__iter__` (1). Audit:

```bash
$ rg "for .* in (\w*da\w*|.*\bDistributionArray)" probpipe/ tests/ docs/
```

Most call sites today either:
- iterate a 1-D `DistributionArray` (unaffected — leading-axis and flat are the same), or
- already use `da.components` / `da[i]` / `_flat_component` for explicit positional access.

Tests that rely on flat iteration over multi-d arrays migrate to `da.components` (still flat). The new behaviour is observable for users who wrote `list(da)` against a multi-d backend; they get sub-arrays now, matching what they'd get from `list(np.zeros(da.batch_shape))`.

`.size` (2) and the doc fix (3) are additive / cosmetic.

## Suggested resolution

Single-PR cleanup, sequenced after PR-C.3 (#176) lands. Touches:

- `probpipe/core/_distribution_array.py` — `__iter__` rewrite + `size` property + module-docstring polish.
- `probpipe/core/_distribution_array.py:67-68` — drop the leftover `n` reference in the class-docstring container surface tuple (already flagged in #176's review pass).
- `CONTRIBUTING.md:391-392` — fix the `len(da)` formula.
- `STYLE_GUIDE.md §1.11` — extend the existing "len/prod" sentence to mention `da.size`.
- Tests covering: 1-D iteration unchanged, multi-d iteration yields sub-arrays of correct shape, `size == prod(batch_shape)`, sub-array iteration matches `np.ndarray` reference behaviour.

## References

- `probpipe/core/_distribution_array.py:426-427` — `__len__` (leading-axis, correct)
- `probpipe/core/_distribution_array.py:498-501` — `__iter__` (flat, to change)
- numpy: `len(np.zeros((3, 4))) == 3`, `np.zeros((3, 4)).size == 12`
- jax: same conventions as numpy
- PR-C.3 (#176) — surfaced these gaps when codifying the `len` / `prod` distinction in `STYLE_GUIDE.md` §1.11

**Comment by jhuggins** [comment]
Done in PR-C.3 (#176). `DistributionArray.__iter__` walks the leading axis (1-D yields scalar cells, multi-d yields sub-arrays of shape `batch_shape[1:]`, 0-d raises `TypeError` like numpy); `DistributionArray.size` returns `prod(batch_shape)`; STYLE_GUIDE §1.11 and the CONTRIBUTING.md framework-hierarchy rule both updated to match.
